enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageDecoder.java

```diff
@@ -0,0 +1,230 @@
+ * under separate terms from the AGPL. Inquiries can be directed to:
+ * licensing@neo4j.com
+ *
+ * More information is also available at:
+ * https://neo4j.com/licensing/
+ */
+package org.neo4j.causalclustering.messaging.marshalling.v2.decoding;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+import java.io.IOException;
+import java.time.Clock;
+import java.util.List;
+import java.util.Queue;
+import java.util.function.BiFunction;
+
+import org.neo4j.causalclustering.catchup.Protocol;
+import org.neo4j.causalclustering.core.consensus.RaftMessages;
+import org.neo4j.causalclustering.core.consensus.log.RaftLogEntry;
+import org.neo4j.causalclustering.core.replication.ReplicatedContent;
+import org.neo4j.causalclustering.identity.ClusterId;
+import org.neo4j.causalclustering.identity.MemberId;
+import org.neo4j.causalclustering.messaging.EndOfStreamException;
+import org.neo4j.causalclustering.messaging.NetworkReadableClosableChannelNetty4;
+import org.neo4j.causalclustering.messaging.marshalling.v2.ContentType;
+import org.neo4j.storageengine.api.ReadableChannel;
+
+import static org.neo4j.causalclustering.core.consensus.RaftMessages.Type.APPEND_ENTRIES_REQUEST;
+import static org.neo4j.causalclustering.core.consensus.RaftMessages.Type.APPEND_ENTRIES_RESPONSE;
+import static org.neo4j.causalclustering.core.consensus.RaftMessages.Type.HEARTBEAT;
+import static org.neo4j.causalclustering.core.consensus.RaftMessages.Type.HEARTBEAT_RESPONSE;
+import static org.neo4j.causalclustering.core.consensus.RaftMessages.Type.LOG_COMPACTION_INFO;
+import static org.neo4j.causalclustering.core.consensus.RaftMessages.Type.NEW_ENTRY_REQUEST;
+import static org.neo4j.causalclustering.core.consensus.RaftMessages.Type.PRE_VOTE_REQUEST;
+import static org.neo4j.causalclustering.core.consensus.RaftMessages.Type.PRE_VOTE_RESPONSE;
+import static org.neo4j.causalclustering.core.consensus.RaftMessages.Type.VOTE_REQUEST;
+import static org.neo4j.causalclustering.core.consensus.RaftMessages.Type.VOTE_RESPONSE;
+
+public class RaftMessageDecoder extends ByteToMessageDecoder
+{
+    private final Protocol<ContentType> protocol;
+
+    RaftMessageDecoder( Protocol<ContentType> protocol )
+    {
+        this.protocol = protocol;
+    }
+
+    @Override
+    public void decode( ChannelHandlerContext ctx, ByteBuf buffer, List<Object> list ) throws Exception
+    {
+        ReadableChannel channel = new NetworkReadableClosableChannelNetty4( buffer );
+        ClusterId clusterId = ClusterId.Marshal.INSTANCE.unmarshal( channel );
+
+        int messageTypeWire = channel.getInt();
+        RaftMessages.Type[] values = RaftMessages.Type.values();
+        RaftMessages.Type messageType = values[messageTypeWire];
+
+        MemberId from = retrieveMember( channel );
+        BiFunction<Queue<Long>,Queue<ReplicatedContent>,RaftMessages.BaseRaftMessage> result;
```

Comment:
I think it would be good to extract this BiFunction into a proper interface of its own to add some semantic meaning to it. Should probably live somewhere close to RaftMessageCreator.

I think the output type should be RaftMessage and not BaseRaftMessage.

Reply:
I think I understand what you mean. I'll show you a suggestion when I have one.